File: backend/app/models/movements.py
import enum
from sqlalchemy import String, Enum, CheckConstraint, ForeignKey, Integer
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base, TimestampMixin


class MovementType(enum.Enum):
    inbound = "inbound"
    transfer = "transfer"
    to_cabinet = "to_cabinet"
    from_cabinet = "from_cabinet"
    direct_to_cabinet = "direct_to_cabinet"
    writeoff = "writeoff"
    adjustment = "adjustment"


class EquipmentMovement(Base, TimestampMixin):
    __tablename__ = "equipment_movements"

    id: Mapped[int] = mapped_column(primary_key=True)
    movement_type: Mapped[MovementType] = mapped_column(
        Enum(MovementType, name="movement_type"), nullable=False, index=True
    )

    equipment_type_id: Mapped[int] = mapped_column(
        ForeignKey("equipment_types.id"), index=True, nullable=False
    )
    quantity: Mapped[int] = mapped_column(Integer, nullable=False)

    from_warehouse_id: Mapped[int | None] = mapped_column(ForeignKey("warehouses.id"), index=True)
    to_warehouse_id: Mapped[int | None] = mapped_column(ForeignKey("warehouses.id"), index=True)

    from_cabinet_id: Mapped[int | None] = mapped_column(ForeignKey("cabinets.id"), index=True)
    to_cabinet_id: Mapped[int | None] = mapped_column(ForeignKey("cabinets.id"), index=True)

    reference: Mapped[str | None] = mapped_column(String(200))
    comment: Mapped[str | None] = mapped_column(String(1000))

    performed_by_id: Mapped[int] = mapped_column(ForeignKey("users.id"), index=True, nullable=False)

    __table_args__ = (
        CheckConstraint("quantity > 0", name="ck_equipment_movements_qty_positive"),
    )

File: frontend/src/api/movements.ts
import { apiFetch } from "./client";

export type BatchDirectToCabinetItem = {
  equipment_type_id: number;
  quantity: number;
  reference?: string | null;
  comment?: string | null;
};

export type MovementRecord = {
  id: number;
  movement_type: string;
  equipment_type_id: number;
  quantity: number;
  from_warehouse_id?: number | null;
  to_warehouse_id?: number | null;
  from_cabinet_id?: number | null;
  to_cabinet_id?: number | null;
  to_assembly_id?: number | null;
  reference?: string | null;
  comment?: string | null;
  performed_by_id: number;
};

export function createDirectToCabinetBatch(payload: {
  movement_type: "direct_to_cabinet";
  to_cabinet_id: number;
  items: BatchDirectToCabinetItem[];
}) {
  return apiFetch<MovementRecord[]>("/movements/batch", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}
